cpp/inc/bond/core/capped_allocator.h

```diff
@@ -0,0 +1,680 @@
+// Copyright (c) Microsoft. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+#pragma once
+
+#include "config.h"
+#include <boost/optional.hpp>
+#include <boost/static_assert.hpp>
+#include <boost/utility/enable_if.hpp>
+#include <boost/core/ref.hpp>
+#include <boost/intrusive_ptr.hpp>
+#include <atomic>
+#include <functional>
+
+
+namespace bond
+{
+    /// @brief Non-sharable counter to be used with \ref capped_allocator.
+    ///
+    /// @tparam T underlying counter type.
+    ///
+    /// @tparam SingleThreaded flag to indicate thread safety.
+    template <typename T = std::size_t, bool SingleThreaded = false>
+    class capped_allocator_counter;
+
+    /// @brief Shared counter to be used with \ref capped_allocator.
+    ///
+    /// @tparam T underlying counter type.
+    ///
+    /// @tparam SingleThreaded flag to indicate thread safety.
+    template <typename T = std::size_t, bool SingleThreaded = false>
+    class capped_allocator_shared_counter;
+
+    /// @brief STL-compatible allocator adapter that limits the allocations.
+    ///
+    /// @tparam Alloc underlying allocator type.
+    ///
+    /// @tparam Counter underlying counter type.
+    ///
+    /// @remarks The provided counter is used to measure allocations in bytes.
+    template <typename Alloc, typename Counter = capped_allocator_shared_counter<typename std::allocator_traits<Alloc>::size_type>>
+    class capped_allocator;
+
+} // namespace bond
+
+
+namespace std
+{
+    template <typename T, bool SingleThreaded, typename Alloc>
+    struct uses_allocator<bond::capped_allocator_shared_counter<T, SingleThreaded>, Alloc> : std::true_type
+    {};
+
+} // namespace std
+
+
+namespace bond
+{
+    namespace detail
+    {
+        /// @brief Helper type that can hold either a value or a reference.
```

Comment:
Consider moving to separate file in detail folder.

Reply:
Ok, will split the types into separate files.